File: android/app/src/main/java/com/ayuvo/health/export/HealthExportFormat.kt
package com.ayuvo.health.export

import com.ayuvo.health.data.health.HealthDayKeys
import com.ayuvo.health.data.health.HealthSampleRow
import com.ayuvo.health.data.health.HealthSeriesPoint
import com.ayuvo.health.data.health.HealthSourceRow
import com.ayuvo.health.models.HealthDataType
import com.ayuvo.health.models.HealthDayAttribution
import kotlinx.serialization.EncodeDefault
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.MapSerializer
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import java.time.Instant
import java.time.OffsetDateTime
import java.time.ZoneId
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter

/**
 * `ayuvo-health-data` v1 — the platform-neutral zip both apps read and write
 * (docs/health-data-export.md). Entry order lets a single-pass reader preview:
 * manifest.json → samples.ndjson → series.ndjson (optional) → sources.json → rollups.ndjson
 * (optional) → checksums.json. kotlinx.serialization (existing `$$serializer` keep rules).
 */
object HealthExportFormat {
    const val FORMAT = "ayuvo-health-data"
    const val VERSION = 1
    const val PERCENT_CONVENTION = "0-100"
    const val MAX_FILE_BYTES = 512L * 1024 * 1024
    const val MAX_LINE_BYTES = 64 * 1024

    const val ENTRY_MANIFEST = "manifest.json"
    const val ENTRY_SAMPLES = "samples.ndjson"
    const val ENTRY_SERIES = "series.ndjson"
    const val ENTRY_SOURCES = "sources.json"
    const val ENTRY_ROLLUPS = "rollups.ndjson"
    const val ENTRY_CHECKSUMS = "checksums.json"

    val json: Json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = false
        explicitNulls = false
        isLenient = true
    }

    val checksumsSerializer = MapSerializer(String.serializer(), String.serializer())

    private val isoFormatter: DateTimeFormatter = DateTimeFormatter.ISO_OFFSET_DATE_TIME

    /** `start_ms + start_offset_s` → ISO-8601 with offset and milliseconds. */
    fun isoWithOffset(ms: Long, offsetS: Int?, zone: ZoneId): String {
        val instant = Instant.ofEpochMilli(ms)
        val offset = offsetS?.let { ZoneOffset.ofTotalSeconds(it) } ?: zone.rules.getOffset(instant)
        return instant.atOffset(offset).format(isoFormatter)
    }

    /** ISO-8601 with offset → epoch ms + offset seconds. Throws on malformed input. */
    fun parseIso(value: String): Pair<Long, Int> {
        val odt = OffsetDateTime.parse(value.trim(), isoFormatter)
        return odt.toInstant().toEpochMilli() to odt.offset.totalSeconds
    }

    fun sampleFromRow(row: HealthSampleRow, sourceName: String?, zone: ZoneId): HealthExportSample = HealthExportSample(
        id = row.id,
        type_id = row.typeId,
        start = isoWithOffset(row.startMs, row.startOffsetS, zone),
        end = isoWithOffset(row.endMs, row.endOffsetS, zone),
        updated = isoWithOffset(row.updatedMs, null, ZoneOffset.UTC),
        value = row.value,
        value2 = row.value2,
        value3 = row.value3,
        value_text = row.valueText,
        unit = row.unit,
        category_value = row.categoryValue,
        title = row.title,
        extra = row.extraJson?.let { runCatching { json.parseToJsonElement(it) as? JsonObject }.getOrNull() },
        count = row.count,
        source_id = row.sourceId,
        source_name = sourceName,
        device = row.device,
        device_type = row.deviceType,
        recording_method = row.recordingMethod,
        client_record_id = row.clientRecordId,
        origin = row.origin
    )

    /** Import: `local_day` is recomputed from start/end + the type's day attribution. */
    fun rowFromSample(sample: HealthExportSample, zone: ZoneId, origin: Int = HealthSampleRow.ORIGIN_IMPORT): HealthSampleRow {
        val (startMs, startOff) = parseIso(sample.start)
        val (endMs, endOff) = parseIso(sample.end)
        val updatedMs = runCatching { parseIso(sample.updated).first }.getOrDefault(endMs)
        val attribution = HealthDataType.byId(sample.type_id)?.dayAttribution ?: HealthDayAttribution.START
        return HealthSampleRow(
            id = sample.id,
            typeId = sample.type_id,
            startMs = startMs,
            endMs = endMs,
            startOffsetS = startOff,
            endOffsetS = endOff,
            localDay = HealthDayKeys.localDay(attribution, startMs, endMs, startOff, endOff, zone),
            value = sample.value,
            value2 = sample.value2,
            value3 = sample.value3,
            valueText = sample.value_text,
            unit = sample.unit,
            categoryValue = sample.category_value,
            title = sample.title,
            extraJson = sample.extra?.toString(),
            count = sample.count.coerceAtLeast(1),
            sourceId = sample.source_id,
            device = sample.device,
            deviceType = sample.device_type,
            recordingMethod = sample.recording_method,
            clientRecordId = sample.client_record_id,
            origin = origin,
            updatedMs = updatedMs
        )
    }

    fun seriesFromPoint(point: HealthSeriesPoint) = HealthExportSeriesPoint(s = point.sampleId, t = point.tMs, v = point.value)

    fun sourceFromRow(row: HealthSourceRow) = HealthExportSource(row.id, row.name, row.deviceModel, row.deviceType, row.lastSeenMs)

    fun encodeSample(sample: HealthExportSample): String = json.encodeToString(HealthExportSample.serializer(), sample)
    fun decodeSample(line: String): HealthExportSample = json.decodeFromString(HealthExportSample.serializer(), line)
    fun encodeSeries(point: HealthExportSeriesPoint): String = json.encodeToString(HealthExportSeriesPoint.serializer(), point)
    fun decodeSeries(line: String): HealthExportSeriesPoint = json.decodeFromString(HealthExportSeriesPoint.serializer(), line)
}

@Serializable
data class HealthExportDateRange(val start: String, val end: String)

/** `record_count` / `series_count` are always written: iOS decodes them as required integers. */
@OptIn(ExperimentalSerializationApi::class)
@Serializable
data class HealthExportTypeInfo(
    val type: String,
    val category: String,
    val kind: String,
    val aggregation: String,
    val unit: String,
    val display_name: String? = null,
    val native_id: String? = null,
    @EncodeDefault val record_count: Long = 0,
    @EncodeDefault val series_count: Long = 0
)

/**
 * The manifest always carries every contract field (iOS decodes `registry_version`,
 * `percent_convention` and `types` as required), even though the JSON config skips defaults
 * for the per-row entries.
 */
@OptIn(ExperimentalSerializationApi::class)
@Serializable
data class HealthExportManifest(
    val format: String,
    val format_version: Int,
    val platform: String,
    val app_version: String,
    val exported_at: String,
    val zone_id: String,
    val date_range: HealthExportDateRange? = null,
    @EncodeDefault val registry_version: Int = HealthDataType.REGISTRY_VERSION,
    @EncodeDefault val percent_convention: String = HealthExportFormat.PERCENT_CONVENTION,
    @EncodeDefault val types: List<HealthExportTypeInfo> = emptyList()
)

@Serializable
data class HealthExportSample(
    val id: String,
    val type_id: String,
    val start: String,
    val end: String,
    val updated: String,
    val value: Double? = null,
    val value2: Double? = null,
    val value3: Double? = null,
    val value_text: String? = null,
    val unit: String,
    val category_value: Int? = null,
    val title: String? = null,
    val extra: JsonObject? = null,
    val count: Int = 1,
    val source_id: String,
    val source_name: String? = null,
    val device: String? = null,
    val device_type: Int? = null,
    val recording_method: Int? = null,
    val client_record_id: String? = null,
    val origin: Int = 0
)

@Serializable
data class HealthExportSeriesPoint(val s: String, val t: Long, val v: Double)

@Serializable
data class HealthExportSource(
    val id: String,
    val name: String,
    val device_model: String? = null,
    val device_type: Int? = null,
    val last_seen_ms: Long? = null
)

File: ios/calorietracker/Services/HealthData/Export/HealthExportFormat.swift
import Foundation

/// `ayuvo-health-data` v1 (`docs/health-data-export.md`). Entry order lets a single-pass
/// reader preview before touching samples; `checksums.json` is last.
nonisolated enum HealthExportFormat {
    static let format = "ayuvo-health-data"
    static let formatVersion = 1
    static let percentConvention = "0-100"

    static let manifestEntry = "manifest.json"
    static let samplesEntry = "samples.ndjson"
    static let seriesEntry = "series.ndjson"
    static let sourcesEntry = "sources.json"
    static let rollupsEntry = "rollups.ndjson"
    static let checksumsEntry = "checksums.json"

    static let maxArchiveBytes = 512 * 1024 * 1024
    static let maxLineBytes = 64 * 1024
    /// Intake is covered by the food-diary export; hydration is still exported.
    static let excludedTypePrefixes = ["dietary_"]
    static let excludedTypes: Set<String> = ["nutrition"]

    static func isExportable(typeID: String) -> Bool {
        !excludedTypes.contains(typeID) && !excludedTypePrefixes.contains { typeID.hasPrefix($0) }
    }

    struct DateRange: Codable, Equatable, Sendable {
        var start: String?
        var end: String?
    }

    struct TypeEntry: Codable, Equatable, Sendable {
        var type: String
        var category: String
        var kind: String
        var aggregation: String
        var unit: String
        var display_name: String?
        var native_id: String?
        var record_count: Int
        var series_count: Int
    }

    struct Manifest: Codable, Equatable, Sendable {
        var format: String
        var format_version: Int
        var platform: String
        var app_version: String
        var exported_at: String
        var zone_id: String
        var date_range: DateRange
        var registry_version: Int
        var percent_convention: String
        var types: [TypeEntry]
    }

    struct SourceEntry: Codable, Equatable, Sendable {
        var id: String
        var name: String
        var device_model: String?
        var device_type: Int?
        var last_seen: String?
    }

    /// One `samples.ndjson` line from a stored row (`local_day` is not exported; importers recompute it).
    static func sampleLine(_ row: HealthSampleRow, sourceName: String?, deviceZoneOffsetS: Int) -> [String: Any] {
        let startOffset = row.startOffsetS ?? deviceZoneOffsetS
        let endOffset = row.endOffsetS ?? startOffset
        var extra: Any = NSNull()
        if let extraJSON = row.extraJSON, let data = extraJSON.data(using: .utf8),
           let object = try? JSONSerialization.jsonObject(with: data) {
            extra = object
        }
        return [
            "id": row.id,
            "type_id": row.typeID,
            "start": ISO8601Fast.format(ms: row.startMs, offsetS: startOffset),
            "end": ISO8601Fast.format(ms: row.endMs, offsetS: endOffset),
            "updated": ISO8601Fast.format(ms: row.updatedMs, offsetS: 0),
            "value": row.value as Any? ?? NSNull(),
            "value2": row.value2 as Any? ?? NSNull(),
            "value3": row.value3 as Any? ?? NSNull(),
            "value_text": row.valueText as Any? ?? NSNull(),
            "unit": row.unit,
            "category_value": row.categoryValue as Any? ?? NSNull(),
            "title": row.title as Any? ?? NSNull(),
            "extra": extra,
            "count": row.count,
            "source_id": row.sourceID,
            "source_name": sourceName as Any? ?? NSNull(),
            "device": row.device as Any? ?? NSNull(),
            "device_type": row.deviceType as Any? ?? NSNull(),
            "recording_method": row.recordingMethod as Any? ?? NSNull(),
            "client_record_id": row.clientRecordID as Any? ?? NSNull(),
            "origin": row.origin,
        ]
    }

    /// Parses one line back into a row. `type` supplies day attribution and the canonical
    /// unit check; returns nil for structurally invalid lines.
    static func row(fromLine object: [String: Any], type: HealthMetricType, calendar: Calendar) -> HealthSampleRow? {
        guard let id = object["id"] as? String, !id.isEmpty,
              let typeID = object["type_id"] as? String, !typeID.isEmpty,
              let startText = object["start"] as? String, let start = ISO8601Fast.parse(startText),
              let endText = object["end"] as? String, let end = ISO8601Fast.parse(endText),
              let unit = object["unit"] as? String
        else { return nil }
        let updated = (object["updated"] as? String).flatMap(ISO8601Fast.parse)?.ms ?? end.ms
        var extraJSON: String?
        if let extra = object["extra"] as? [String: Any], !extra.isEmpty {
            extraJSON = HealthSampleMapper.json(extra)
        }
        return HealthSampleRow(
            id: id,
            typeID: typeID,
            startMs: start.ms,
            endMs: max(end.ms, start.ms),
            startOffsetS: start.offsetS,
            endOffsetS: end.offsetS,
            localDay: HealthRollupMath.localDay(
                startMs: start.ms, endMs: max(end.ms, start.ms), startOffsetS: start.offsetS, endOffsetS: end.offsetS,
                attribution: type.dayAttribution, calendar: calendar
            ),
            value: number(object["value"]),
            value2: number(object["value2"]),
            value3: number(object["value3"]),
            valueText: object["value_text"] as? String,
            unit: unit,
            categoryValue: number(object["category_value"]).map { Int($0) },
            title: object["title"] as? String,
            extraJSON: extraJSON,
            count: max(1, number(object["count"]).map { Int($0) } ?? 1),
            sourceID: (object["source_id"] as? String) ?? "unknown",
            device: object["device"] as? String,
            deviceType: number(object["device_type"]).map { Int($0) },
            recordingMethod: number(object["recording_method"]).map { Int($0) },
            clientRecordID: object["client_record_id"] as? String,
            origin: HealthRowOrigin.fileImport.rawValue,
            deleted: 0,
            updatedMs: updated
        )
    }

    static func number(_ value: Any?) -> Double? {
        switch value {
        case let n as NSNumber: return n.doubleValue
        case let d as Double: return d
        case let i as Int: return Double(i)
        case let s as String: return Double(s)
        default: return nil
        }
    }
}
